controllers/managedocs_controller.go

```diff
@@ -551,25 +572,78 @@ func (r *ManagedOCSReconciler) reconcileAlertmanager() error {
-	// _, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfig, func() error {
-	// 	if err := r.own(r.alertmanagerConfig); err != nil {
-	// 		return err
-	// 	}
+func (r *ManagedOCSReconciler) reconcileAlertmanagerConfigSecret() error {
+	r.Log.Info("Reconciling AlertmanagerConfig secret")
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfigSecret, func() error {
+		if err := r.own(r.alertmanagerConfigSecret); err != nil {
+			return err
+		}
+
+		if err := r.get(r.pagerdutySecret); err != nil {
+			return fmt.Errorf("Unable to get pagerduty secret: %v", err)
+		}
 
-	// 	desired := utils.ObjectFromTemplate(templates.AlertmanagerConfigTemplate, r.Scheme).(*promv1a1.AlertmanagerConfig)
-	// 	r.alertmanagerConfig.ObjectMeta.Labels = map[string]string{monLabelKey: monLabelValue}
-	// 	// TODO: Setup pagerduty and deadman sqtich on the desired alertmanagr config
-	// 	r.alertmanagerConfig.Spec = desired.Spec
+		pagerdutySecretData := r.pagerdutySecret.Data
+		pagerdutyServiceKey := string(pagerdutySecretData["PAGERDUTY_KEY"])
+		if pagerdutyServiceKey == "" {
+			pagerdutyServiceKey = string(pagerdutySecretData["serviceKey"])
+		}
+		if pagerdutyServiceKey == "" {
+			return fmt.Errorf("Pagerduty secret does not contain a serviceKey or PAGERDUTY_KEY entries")
+		}
 
-	// 	return nil
-	// })
-	// if err != nil {
-	// 	return err
-	// }
+		config, err := r.getAlertmanagerConfig(pagerdutyServiceKey)
+		if err != nil {
+			return err
+		}
+		r.alertmanagerConfigSecret.Data = map[string][]byte{
+			"alertmanager.yaml": config,
+		}
+		return nil
+	})
 
-	return nil
+	return err
+}
+
+func (r *ManagedOCSReconciler) getAlertmanagerConfig(pagerdutyServiceKey string) ([]byte, error) {
+
+	alertmanagerConfig := struct {
+		Route struct {
+			GroupWait      string `yaml:"group_wait"`
+			GroupInterval  string `yaml:"group_interval"`
+			RepeatInterval string `yaml:"repeat_interval"`
+			Receiver       string `yaml:"receiver"`
+			Routes         [1]struct {
+				Match struct {
+					Severity string `yaml:"severity"`
+				} `yaml:"match"`
+				Receiver string `yaml:"receiver"`
+			} `yaml:"routes"`
+		} `yaml:"route"`
+		Receivers [1]struct {
+			Name             string `yaml:"name"`
+			PagerdutyConfigs [1]struct {
+				ServiceKey string `yaml:"service_key"`
+			} `yaml:"pagerduty_configs"`
+		} `yaml:"receivers"`
+	}{}
+
+	alertmanagerConfig.Route.GroupWait = "30s"
+	alertmanagerConfig.Route.GroupInterval = "5m"
+	alertmanagerConfig.Route.RepeatInterval = "4h"
+	alertmanagerConfig.Route.Receiver = "pagerduty"
+	alertmanagerConfig.Route.Routes[0].Match.Severity = "critical"
+	alertmanagerConfig.Route.Routes[0].Receiver = "pagerduty"
+	alertmanagerConfig.Receivers[0].Name = "pagerduty"
+	alertmanagerConfig.Receivers[0].PagerdutyConfigs[0].ServiceKey = pagerdutyServiceKey
+
+	config, err := yaml.Marshal(alertmanagerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse alertmanager conifg: %v", err)
+	}
```

Comment:
I think this conversion code should be located outside this function, as the original request was to separate the acquisition of the config from the processing and the reconciliation of it.
Also, the cadence of change for the data itself and for the mechanism of processing it. is not the same at all